### ISSUE TITLE
connect to local SQL server

### ISSUE THREAD

**Issue by ophiryaniv-ts** [GoogleCloudPlatform/cloud-sql-python-connector#856]
### Question

Hey,
What is the best practice for working with `cloud-sql-python-connector` with local SQL.
I like to create a local SQL server when developing and use the GCP Manage SQL in DEV/STG/PROD.

But using `localhost` as `instance_connection_string` gets me `ValueError`.
So I wonder what should I do?

**Comment by jackwotherspoon** [comment]
Hi @ophiryaniv-ts, thanks for raising an issue on the Cloud SQL Python Connector! 😄 

The Cloud SQL Python Connector is designed purely for connecting to remote Cloud SQL databases, it is not designed to connect to a local database.

If you want to connect to a local SQL database I'd recommend doing so with the [SQLAlchemy](https://docs.sqlalchemy.org/en/20/core/engines.html#mysql) library which is also the recommended usage of the Cloud SQL Python Connector. This will allow you to have very similar implementation for DEV/STG/PROD.

A second option would be to leverage a Private IP Cloud SQL connection and connect in STG/PROD directly to your Cloud SQL instance's Private IP address without the Cloud SQL Python Connector. This would allow you to have the same code for DEV/STG/PROD.

Basic code sample using username and password auth:
```python
import sqlalchemy
from sqlalchemy import event

engine = sqlalchemy.create_engine(
    # Equivalent URL:
    # postgresql+psycopg2://<user>:<password>@<host>:5432/<db_name>
    sqlalchemy.engine.url.URL.create(
        drivername="postgresql+psycopg2",
        username=user,  #  "my-user"
        password="",  # "my-password"
        host=ip_address,  # Cloud SQL instance IP address (or localhost for local SQL)
        port=5432,  # 5432 for Postgres, 3306 for MySQL
        database=db_name,  # "my-database-name"
    ),
    connect_args={"sslmode": "require"},
)

# use connection from connection pool to query AlloyDB database
with engine.connect() as conn:
    time = conn.execute(sqlalchemy.text("SELECT NOW()")).fetchone()
    conn.commit()
    print("Current time is ", time[0])
```

If you are leveraging [Automatic IAM Database Authentication](https://cloud.google.com/sql/docs/postgres/authentication#automatic) you can continue to do so even without the Connector:
```python
import sqlalchemy
from sqlalchemy import event

import google.auth
from google.auth.credentials import Credentials
from google.auth.transport.requests import Request

# initialize Google Auth creds
creds, _ = google.auth.default(
    scopes=["https://www.googleapis.com/auth/cloud-platform"]
)

def get_authentication_token(credentials: Credentials) -> str:
    """Get OAuth2 access token to be used for IAM database authentication"""
    # refresh credentials if expired
    if not credentials.valid:
        request = Request()
        credentials.refresh(request)
    return credentials.token

engine = sqlalchemy.create_engine(
    # Equivalent URL:
    # postgresql+psycopg2://<user>:empty@<host>:5432/<db_name>
    sqlalchemy.engine.url.URL.create(
        drivername="postgresql+psycopg2",
        username=user,  # IAM db user, e.g. service-account@project-id.iam
        password="",  # placeholder to be replaced with OAuth2 token
        host=ip_address,  # Cloud SQL instance IP address
        port=5432,
        database=db_name,  # "my-database-name"
    ),
    connect_args={"sslmode": "require"},
)

# set 'do_connect' event listener to replace password with OAuth2 token
@event.listens_for(engine, "do_connect")
def auto_iam_authentication(dialect, conn_rec, cargs, cparams) -> None:
    cparams["password"] = get_authentication_token(creds)

# use connection from connection pool to query AlloyDB database
with engine.connect() as conn:
    time = conn.execute(sqlalchemy.text("SELECT NOW()")).fetchone()
    conn.commit()
    print("Current time is ", time[0])
```

Hope this helps. Thanks again for the question!

**Comment by ophiryaniv-ts** [comment]
Hey, @jackwotherspoon thanks for the answer :) 

I am using Postgresql DB with `asyncpg` as my driver. 
I can use you answer and leverage the Private IP Cloud SQL so I will get:
```python
engine = await asyncpg.connect(
  user=user,  #  "my-user"
  password="",  # "my-password"
  host=ip_address,  # Cloud SQL instance IP address (or localhost for local SQL)
  port=5432,  # 5432 for Postgres
  database=db_name,  # "my-database-name"
                )
```
But is it really recommend the same as using the connector? 

Second option is to use an if-else:
 ```python
if env=="local":
  engine = await asyncpg.connect(
    user=user,  #  "my-user"
    password="",  # "my-password"
    host=ip_address,  # Cloud SQL instance IP address (or localhost for local SQL)
    port=5432,  # 5432 for Postgres
    database=db_name,  # "my-database-name"
  )
else: 
  engine = await connector.connect_async(...)
```

What do you think?

**Comment by jackwotherspoon** [comment]
Hi @ophiryaniv-ts we always recommend connection pooling, if you want to use native `asyncpg` I would recommend taking advantage of the libraries native connection pooling, otherwise SQLAlchemy also support `asyncpg`.

Native asyncpg lib (also don't forget to require SSL):
```python
import asyncpg 

# ... put this inside of async context (function)
async with asyncpg.create_pool(
    user="my-user",
    password="my-pass",
    host=ip_address,  # your Cloud SQL instance IP address
    port=5432,
    database=db,  # your database name
    ssl="require",
) as pool:
    # acquire connection from native asyncpg connection pool
    async with pool.acquire() as conn:
        time = await conn.fetchrow("SELECT NOW()")
        print("Current time is ", time[0])
```

SQLAlchemy using async engine:
```python
import sqlalchemy
from sqlalchemy.ext.asyncio import create_async_engine

engine = create_async_engine(
      # Equivalent URL:
      # postgresql+asyncpg://<user>:<pass>@<host>:5432/<db_name>
      sqlalchemy.engine.url.URL.create(
          drivername="postgresql+asyncpg",
          username="my-user",
          password="my-pass", 
          host=ip_address,  # your Cloud SQL instance IP address
          port=5432,
          database=db_name,  # your database name
      ),
      connect_args={"ssl": "require"},
  )
```

> But is it really recommend the same as using the connector?

TLDR; The Python Connector is recommended for Public IP because of the automated mTLS connection but for Private IP a direct TCP database connection is secure and more performant.

This depends on your use-case, the Python Connector provides two main benefits:
- support of automated mTLS connections (SSL connection using client certs)
- [Automatic IAM Database Authentication](https://cloud.google.com/sql/docs/postgres/authentication#automatic)

We recommend the connectors primarily for Public IP Cloud SQL connections because of the automated mTLS connections. The connector generates a client certificate and uses it to establish a secure SSL connection. However, client certificates are not a necessity for secure connections. If you are using Private IP then your connection is via a private VPC network, the requirement for client certs is not as prominent in this case and setting `ssl="require"` will still establish a secure SSL connection (which may be more than enough for your use case.

For automatic IAM authentication we have found ways to do this without the connectors as I showed in my previous response. Additional samples showing how to use auto IAM database authentication with asyncpg can be found [here](https://github.com/GoogleCloudPlatform/alloydb-python-connector/blob/main/tests/system/test_native_asyncpg_direct_connection.py) and [here](https://github.com/GoogleCloudPlatform/alloydb-python-connector/blob/main/tests/system/test_sqlalchemy_asyncpg_direct_connection.py)

Additionally the use of the Python Connecotr (or any of the language connectors) adds some addtional network hops to reach your Cloud SQL instance which may introduce some slight performance delays compared to the direct TCP connection using Private IP.
